Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h

```diff
@@ -44,38 +46,51 @@ class TextureMapperLayer3DRenderingContext final {
         const std::function<void(TextureMapperLayer*, const ClipPath&)>&);
 
 private:
-    enum class PolygonPosition {
+    enum class LayerPosition {
         InFront,
         Behind,
         Coplanar,
         Intersecting
     };
 
-    struct TextureMapperLayerPolygon final {
+    enum class SweepAndPruneAxis {
+        XAxis,
+        YAxis,
+        ZAxis
+    };
+
+    struct Layer final {
         FloatPolygon3D geometry;
-        TextureMapperLayer* layer = { nullptr };
-        bool isSplitted = { false };
+        TextureMapperLayer* textureMapperLayer { nullptr };
+        bool isSplitted { false };
         unsigned clipVertexBufferOffset { 0 };
     };
 
-    struct TextureMapperLayerNode final {
+    struct LayerNode final {
         WTF_MAKE_STRUCT_FAST_ALLOCATED;
 
-        explicit TextureMapperLayerNode(TextureMapperLayerPolygon&& polygon)
+        explicit LayerNode(Layer&& layer)
         {
-            polygons.append(WTFMove(polygon));
+            layers.append(WTFMove(layer));
         }
 
-        const TextureMapperLayerPolygon& firstPolygon() const  { return polygons[0]; }
+        const Layer& firstLayer() const  { return layers[0]; }
 
-        Vector<TextureMapperLayerPolygon> polygons;
-        std::unique_ptr<TextureMapperLayerNode> frontNode;
-        std::unique_ptr<TextureMapperLayerNode> backNode;
+        Vector<Layer> layers;
+        std::unique_ptr<LayerNode> frontNode;
+        std::unique_ptr<LayerNode> backNode;
     };
 
-    void buildTree(TextureMapperLayerNode&, Deque<TextureMapperLayerPolygon>&);
-    void traverseTree(TextureMapperLayerNode&, const std::function<void(TextureMapperLayerNode&)>&);
-    static PolygonPosition classifyPolygon(const TextureMapperLayerPolygon&, const FloatPlane3D&);
+    using SweepAndPrunePairs = StdSet<std::pair<size_t, size_t>>;
+
+    SweepAndPrunePairs sweepAndPrune(const Vector<Layer>&, SweepAndPruneAxis) const;
+    FloatQuad projectLayerToZYPlane(const Layer&) const;
+    FloatPoint edgeNormal(const FloatPoint&, const FloatPoint&) const;
+    std::pair<float, float> projectQuadOnAxis(const FloatQuad&, const FloatPoint& axis) const;
+    bool quadsIntersect(const FloatQuad&, const FloatQuad&) const;
+    void buildTree(LayerNode&, Deque<Layer>&);
+    void traverseTree(LayerNode&, const std::function<void(LayerNode&)>&);
+    static LayerPosition classifyLayer(const Layer&, const FloatPlane3D&);
```

Comment:
I think we should add `static` to member functions which don't need `this`.

Reply:
Fixed